cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -173,6 +195,84 @@ protected void addPartition(PartitionKey key, String path, boolean explorable, M
   }
 
   @Override
+  public PartitionConsumerResult consumePartitions(PartitionConsumerState partitionConsumerState) {
+    Set<Long> previousInProgress = partitionConsumerState.getVersionsToCheck();
+    Set<Long> noLongerInProgress = setDiff(previousInProgress, tx.getInProgress());
+
+    List<Iterator<Partition>> partitionIterators = Lists.newArrayList();
+    for (Long txId : noLongerInProgress) {
+      Scanner scanner = partitionsTable.readByIndex(WRITE_PTR_COL, Bytes.toBytes(txId));
+      partitionIterators.add(new PartitionIterator(scanner));
+    }
+
+    long readPointer = tx.getReadPointer();
+    // no read your own writes (partitions)
+    Scanner scanner = partitionsTable.scanByIndex(WRITE_PTR_COL,
+                                                  Bytes.toBytes(partitionConsumerState.getStartVersion()),
+                                                  Bytes.toBytes(readPointer + 1));
+    partitionIterators.add(new PartitionIterator(scanner));
+
+    Set<Long> inProgressBeforeReadPointer = Sets.newHashSet();
+    for (long txId : tx.getInProgress()) {
+      if (txId >= readPointer) {
+        break;
+      }
+      inProgressBeforeReadPointer.add(txId);
+    }
+    return new PartitionConsumerResult(new PartitionConsumerState(readPointer + 1, inProgressBeforeReadPointer),
+                                       Iterators.concat(partitionIterators.iterator()));
+  }
+
+  private Set<Long> setDiff(Set<Long> oldLongsSet, long[] newLongs) {
```

Comment:
not entirely sure whether this does what it should... 
You want all versions that were in oldLongsSet but are not in newLongs. But do you want to alter oldLongSet?
I think you better make a copy of oldLongSet, then remove every element of newLongs from it, and return it.